### DISCUSSION TITLE
Pattern: Form required fields

### DISCUSSION THREAD

**Discussion by mimarz** [digdir/designsystemet#4223] [Patterns for public services]
How do we signify required form fields? 

- Asterisks on label? 
- "required" text after label? 
- Preface that all fields in a form are required?
- If all form fields are required by default, then only mark optional fields?

There are multiple solutions and aspects to this we need to figure out in terms of which form patterns we recommend and what we implement in our components.

Discussion in Slack: https://altinndevops.slack.com/archives/C0510HS2RRR/p1692338419113949

**Comment by Camulos** [comment]
Først de aktuelle reglene:
- WCAG: 1.1.1: Ikke-tekstlig innhold: Gi brukeren et tekstalternativ for innhold som ikke er tekst.
- WCAG: 3.3.2: Ledetekster eller instruksjoner: Det vises ledetekster eller instruksjoner når du har skjemaelementer som må fylles ut.
- Elmer: 3.6 Frivillige felter: Hvis det er frivillig å besvare et spørsmål skal dette framgå tydelig.

Vi har også et WCAG suksesskriterie om at en skal vise til mer enn farge, form etc for å veilede...

I BR har dette resultert i følgende krav til løsning:
Vi forteller med tekst hvilke felter som er påkrevde eller ikke-påkrevde. Det betyr ikke at det skal stå påkrevd på alle påkrevde felt og ikke-påkrevd på alle frivillige felter i samme løsning, men en av delene skal være merket med tekst i nærheten av feltet / label, vanligvis etter eller som del av label.

Dersom en likevel vil bruke stjernetegnet, så skal det også være forklart hva stjernetegnet betyr et sted på hver eneste side det brukes, dette fordi det er det som er poenget med stjernetegnet, å fortelle at her er det en notis som står feks i bunnteksten som sier noe mer om dette temaet eller gir nyttig ekstrainformasjon (tenk en fotnotis).

Så min mening: bruk tekst som del av label og merk enten det ene eller det andre (påkrevd / ikke påkrevd).

**Comment by mrosvik** [comment]
Vi må komme tilbake til å lande felles retningslinjer for hvordan vi skal håndtere dette i ulike situasjoner. 

SSB bør involveres da de har hatt meninger om dette i forbindelse med skjemaer i Altinn Apps. @Febakke sjekker med dem.

**Comment by mrosvik** [comment]
Det jobbes nå med å lande felles mønstre for dette sammen med Nav, Skatteetten, Oslo Origo, Brreg, SSB. https://github.com/digdir/designsystem/issues/844

**Comment by mrosvik** [comment]
Etter 3 arbeidsmøter har vi kommet frem til følgende: https://storefront-pr-1157.dev.designsystemet.no/monstre/obligatoriske-og-valgfrie-felt Artikkelen er sendt til deltakere for tilbakemeldinger.

**Comment by mrosvik** [comment]
Artikkelen kan ses her: https://www.designsystemet.no/no/patterns/required-and-optional-fields

**Comment by mrosvik** [comment]
Det ble gjort endringer på artikkelen 29. januar 2026 basert på tilbakemeldinger. Endringene ble sendt ut til hovedgruppen for mønster-samarbeidet i forkant av publisering. 

Endringene som er gjort er basert på tilbakemeldinger vi fikk:
- Flere reagerer på formuleringen «når man må ha …»
- Noen tolket eksemplene som at man kun skulle ha ett felt per side. Det har ikke vært intensjonen med artikkelen.

Den nyeste versjonen fjerner derfor det første eksempelet som har skapt forvirring. Vi har også justert formuleringene ellers, slik at artikkelen i mindre grad kan tolkes som en anbefaling om hvor mange spørsmål man bør ha per side. Det temaet kan eventuelt få mer fokus i en egen artikkel. 

Forrige versjon av artikkelen kan ses her: https://github.com/digdir/designsystemet/blob/5cad48d1c7c5f891837bfdc2e61cc39b22e6376c/apps/www/app/content/patterns/no/required-and-optional-fields.mdx

**Comment by mrosvik** [comment]
Mulig utvidelse til artikkelen, foreslått av @Camulos :

Should the pattern include this case:
In a form, we have fields A, B, C, and D.

Fields A and B are initially required.
Fields C and D are initially optional.

The question in field B is of the Boolean type (do you want to or not?).
If "true", then field C must also be filled out.
C is therefore no longer optional, and the labeling of the field must be updated.

We can change the tag that indicates whether the field is required or not, but there may be more to this pattern that should be addressed — for example, informing the screen reader about the change,
is that necessary when the change occurs after the field that triggers it?

Opprinnelig publisert i en issue som ble lukket: https://github.com/digdir/designsystemet/issues/3761
